server/src/main/java/org/apache/cloudstack/snapshot/SnapshotHelper.java

```diff
@@ -101,15 +101,21 @@ public void expungeTemporarySnapshot(boolean kvmSnapshotOnlyInPrimaryStorage, Sn
             return;
         }
 
-        logger.debug(String.format("Expunging snapshot [%s] due to it is a temporary backup to create a volume from snapshot. It is occurring because the global setting [%s]"
-          + " has the value [%s].", snapInfo.getId(), SnapshotInfo.BackupSnapshotAfterTakingSnapshot.key(), backupSnapshotAfterTakingSnapshot));
-
-        try {
-            snapshotService.deleteSnapshot(snapInfo);
-        } catch (CloudRuntimeException ex) {
-            logger.warn(String.format("Unable to delete the temporary snapshot [%s] on secondary storage due to [%s]. We still will expunge the database reference, consider"
-              + " manually deleting the file [%s].", snapInfo.getId(), ex.getMessage(), snapInfo.getPath()), ex);
+        if (!DataStoreRole.Image.equals(snapInfo.getDataStore().getRole())) {
```

Comment:
actually still not sure it won't be a potential NPE ; `snapInfo` is not null but `snapInfo.getDataStore()` still may be.

Reply:
Datastore/Role cannot be null based on the existing method calls and it should be null for further steps as well. Do you want me to keep a check anyways ?

Reply:
maybe you are right @harikrishna-patnala (probably actually) but in this public method a snapshotInfo object may be passed that is constructed with a null-datastore. I think an extra check does not hurt here.